### DISCUSSION TITLE
Balancing participants between conditions

### DISCUSSION THREAD

**Discussion by gorkang** [jspsych/jsPsych#1719] [Q&A]
In this [comment](https://github.com/jspsych/jsPsych/discussions/1532#discussioncomment-382051), Josh links to a [library to communicate with a MySQL database](https://github.com/vassar-cogscilab/php-mysql-save-data) and do things like assign subjects to conditions "in a (somewhat) balanced manner".

When working with a server this should work great, but not so much when running experiments in a lab setting locally, where the same computer is used for multiple participants. This is common in places where internet is not very reliable, or when one does not have/wants access to a server.

**Any idea about what would be the best way to assign subjects to conditions in a balanced manner locally/offline?** 

Maybe adapting the above MySQL solution to a set of local plaintext files, or JSON files could work? (having a JSON file for each MySQL table? table_register.json, table_conditions.json and table_data.json)

How are you folks doing this?

Thanks in advance!

**Comment by jodeleeuw** [comment]
Hi @gorkang,

One problem that you'll face is that JavaScript running in the browser can't access (most of) the local file system. So you can't, for example, directly edit a .json file from browser JS, even if you are running the script locally. I believe that the closest you could come is to use something like [IndexedDB](https://developer.mozilla.org/en-US/docs/Web/API/IndexedDB_API/Using_IndexedDB).

The solution that I would recommend is to run a lightweight web server on your local computer. If you use a node.js-based server, for example, then you can access files on the disk through node.js, and your jspsych scripts can communicate with the server via `XMLHttpRequests`. Or you could even just run MySQL locally on your computer and use the scripts I shared above.

**Comment by gorkang** [comment]
Thanks for your response Josh. IndexedDB might be what we were looking for. We will try to adapt your "[library to communicate with a MySQL database](https://github.com/vassar-cogscilab/php-mysql-save-data)" to use IndexedDB in the following weeks. If we manage to get something useful we will share it.

We want the system to be as portable as possible, and sometimes installing something in the local computer is not an option. A typical use case would be going to a school with the experimental protocol in a pen drive folder and just executing an index.html.

**Comment by Franck-El** [comment]
Hi @gorkang,

If I understood well the problem is about balanced conditions.
Maybe you could directly determine in the code "for this participant number: read condition1.json, for this particant number: read condition2.json ... And so on". You can make a switch on the participant number for each condition case. And, if the participant number is equal or higher the number of conditions then you do partNo - nCondition while it is the case. This way if your number of participants at the end of the experiment is a multiple of your number of conditions, each condition have the same number of participants. This could be done easily in js and so it work online or offline.
But if the problem is about having one big file for each condition results at the end I don't see any other options than making a second program to sort the data obtained.

I'm not sure if it answers the question but may be it could help !

Franck

**Reply by gorkang** [reply]
Hi @Franck-El 

Yes, exactly, the issue is how to have a system that can balance participants between conditions. But we need it to work when running the experiment in a local computer without access to internet and without installing anything in the computer (just using the protocol in a pendrive). Hopefully the same system can also work when using a server, although sadly this seems impossible, so a IndexedDB/MySQL hybrid system seems to be the only option.

The issue with a prefixed matching between participant id and conditions is that, we'd want this system to compensate when the experimental death piles up in certain conditions, rebalancing dynamically to end up with a similar number of participants per condition.
